packages/assets-controllers/src/DeFiPositionsController/DeFiPositionsController.ts

```diff
@@ -0,0 +1,247 @@
+  AccountsControllerListAccountsAction;
+
+/**
+ * The external events available to the {@link DeFiPositionsController}.
+ */
+export type DeFiPositionsControllerAllowedEvents =
+  | KeyringControllerUnlockEvent
+  | KeyringControllerLockEvent
+  | TransactionControllerTransactionConfirmedEvent
+  | AccountsControllerAccountAddedEvent;
+
+/**
+ * The messenger of the {@link DeFiPositionsController}.
+ */
+export type DeFiPositionsControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  DeFiPositionsControllerActions | DeFiPositionsControllerAllowedActions,
+  DeFiPositionsControllerEvents | DeFiPositionsControllerAllowedEvents,
+  DeFiPositionsControllerAllowedActions['type'],
+  DeFiPositionsControllerAllowedEvents['type']
+>;
+
+/**
+ * Controller that stores assets and exposes convenience methods
+ */
+export class DeFiPositionsController extends StaticIntervalPollingController()<
+  typeof controllerName,
+  DeFiPositionsControllerState,
+  DeFiPositionsControllerMessenger
+> {
+  readonly #fetchPositions: (
+    accountAddress: string,
+  ) => Promise<DefiPositionResponse[]>;
+
+  readonly #isEnabled: () => boolean;
+
+  /**
+   * DeFiPositionsController constuctor
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - The controller messenger.
+   * @param options.isEnabled - Function that returns whether the controller is enabled. (default: () => true)
+   */
+  constructor({
+    messenger,
+    isEnabled = () => true,
+  }: {
+    messenger: DeFiPositionsControllerMessenger;
+    isEnabled?: () => boolean;
+  }) {
+    super({
+      name: controllerName,
+      metadata: controllerMetadata,
+      messenger,
+      state: getDefaultDefiPositionsControllerState(),
+    });
+
+    this.setIntervalLength(TEN_MINUTES_IN_MS);
+
+    this.#fetchPositions = buildPositionFetcher();
+    this.#isEnabled = isEnabled;
+
+    this.messagingSystem.subscribe('KeyringController:unlock', () => {
+      this.startPolling(null);
+    });
+
+    this.messagingSystem.subscribe('KeyringController:lock', () => {
+      this.stopAllPolling();
+    });
+
+    this.messagingSystem.subscribe(
+      'TransactionController:transactionConfirmed',
+      async (transactionMeta) => {
+        if (!this.#isEnabled()) {
+          return;
+        }
+
+        await this.#updateAccountPositions(transactionMeta.txParams.from);
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAdded',
+      async (account) => {
+        if (!this.#isEnabled() || !account.type.startsWith('eip155:')) {
+          return;
+        }
+
+        await this.#updateAccountPositions(account.address);
+      },
+    );
+  }
+
+  async _executePoll(): Promise<void> {
+    if (!this.#isEnabled()) {
+      return;
+    }
+
+    const accounts = this.messagingSystem.call(
+      'AccountsController:listAccounts',
+    );
+
+    const results = await reduceInBatchesSerially({
+      initialResult: [] as {
```

Comment:
Instead of typecasting this empty array, what do you think about passing a type argument to `reduceInBatchesSerially`?

Reply:
I think that would mean we forfeit the benefits of automatically inferring the types based on the parameters (in this case I would have to specify two types). And that is way worse that asserting IMO.

If you find the assertion unsightly, I can declare the array previously and avoid the assertion.
```
const initialResult: {
      accountAddress: string;
      positions: GroupedDeFiPositionsPerChain | null;
    }[] = [];
```

Reply:
It's okay, this is not bad, I just thought there might be a way to avoid the type assertion.

Reply:
I think there is, but it would involve going to that function and changing the types around a bit. Using the new `NoInfer` on `initialResult`. I just didn't want to start making too many side changes.